expertsystem/data.py

```diff
@@ -207,6 +207,23 @@ def __post_init__(self) -> None:
                 ")"
             )
 
+    def __hash__(self) -> int:
+        return hash(str(self))
+
+    def __repr__(self) -> str:
+        output_string = f"{self.__class__.__name__}("
+        for field in fields(self):
+            value = getattr(self, field.name)
+            if value is None:
+                continue
+            if field.name not in ["mass", "spin", "isospin"] and value == 0:
+                continue
+            if isinstance(value, str):
+                value = f'"{value}"'
+            output_string += f"\n    {field.name}={value},"
+        output_string += "\n)"
+        return output_string
```

Comment:
Why is this needed? If we want some better "string" representation of this for prettier printing, then `__str__` should be overwritten. Btw, I would assume there is a reason why dataclasses list default values on purpose, like guaranteeing same instance even if default values change.

Reply:
It's actually mainly for the Jupyter notebooks. The `Particle` class is an important interface to the outside, so you want it to render nicely when doing for instance:
```python
from expertsystem import io
pdg = io.load_pdg()
pdg["gamma"]
```
in a notebook. So for instance, not have all those `isospin=None` statements.

Note: in this example, Jupyter notebook calls `Particle.__repr__`, not `Particle.__str__`.

Reply:
Also note that he `__repr__` method is checked with `eval` for all instances in the PDG.

Reply:
i think we need `_repr_pretty_` then, see https://stackoverflow.com/questions/41453624/tell-ipython-to-use-an-objects-str-instead-of-repr-for-output

Reply:
Ok I tried it out a bit, but can't get it to work nicely (especially when printing `ParticleCollection`).

Reply:
Apart from that, this PR was originally motivated by my attempts to create some widgets through Sphinx. I'm currently experimenting with [`sphinx-thebe`](https://sphinx-thebe.readthedocs.io/en/latest/use.html), which uses Python instead of Python (that is, you need a `print` statement to render output).

Reply:
Don't know, just seems fishy to me, as in it is a misuse of `__repr__`. If it would be the purpose to serialize objects via `__repr__` and default arguments are not shown, then when the default arguments change backwards compatibility is lost. Not sure if this is the purpose of `__repr__` though

Reply:
It may indeed be fishy, depends on if we understand Python's intention correctly. Still, an `eval` test is performed for all instances in the particle database from the PDG, so the implementation is quite safe (also with regard to default arguments).